doc/rfcs/2025-11-30--InterProcessCommunication.md

```diff
@@ -0,0 +1,440 @@
+"discover" clients when the client first contacts it, usually via the
+Synchronous Mailbox. An alternative design where any process could discover and
+reach out to any other process is possible, but is not the current focus. Note
+that this does not restrain communication patterns after initiation. Both
+client requests and asynchronous server notifications are supported in
+communication mechanisms.
+
+
+## IPC Name Registry Capsule
+
+This capsule provides registration for services and discovery for clients.
+Upon discovery, it provides opaque process IDs which can be used to refer to
+processes in other IPC mechanisms.
+
+If validation of services or clients is desired, this capsule could perform
+that operation at registration-time. The initial design will likely skip that
+feature but the design should enable it to be possible and provide clear
+locations where it could later be connected if implemented. Callbacks will be
+given on registration and discovery, allowing asynchronous operations to take
+place before they are completed.
+
+Registration and discovery is performed by matching "names", which are
+twenty-byte arrays. Allowed buffers must be exactly twenty bytes in length
+or commands using them will fail. The byte values can be anything, but are
+typically ASCII values, which do not need to be null-terminated. The default
+name value of all-zero-values cannot be used for discovery. Values will be
+copied from the allowed buffer into a fixed-size allocation in the grant
+region.
+
+An alternative mechanism of using package identifiers was considered, but was
+less flexible than having arbitrary process-specified identifiers. Package
+identifiers have some advantages that they are 1) arbitrary length, and 2) are
+unable to be modified at runtime. A downside of package identifiers is that
+they are encoded as part of the build system, and not part of the application
+code.
+
+**Commands**:
+* Existence
+* Register as service with allowed string name
+* Discover service with allowed string name
+
+**Allows**:
+* Read-only, string name of service
+
+**Subscribes**:
+* Registration complete (success or failure)
+* Discovery complete (provides process ID)
+
+
+## Synchronous Mailbox Capsule
+
+This capsule provides synchronous client-to-server request-and-response
+messages. Clients can send a request to any process ID. After receiving a
+request message, the server performs whatever action it chooses and eventually
+sends a response message back to the client. Clients may end up waiting
+for an arbitrary duration until the server responds to them, but may also
+cancel their request at any time.
+
+Servers do not need to be aware of clients in advance, as they will receive a
+process ID with each request. Clients do need to have previously discovered a
+process ID for the server, possibly via the IPC Name Registry or possibly via a
+separate mechanism (for example, they could be hard-coded).
```

Comment:
So presumably something in the kernel has to store the list of discovered process IDs for each process, correct?